lib/sedna/service/server/knowledgeBase/server.py

```diff
@@ -0,0 +1,235 @@
+                    session.add(grp)
+                grp.sample_num += len(task_group.samples)
+                grp.task_num += len(task_group.tasks)
+                t_id = []
+                for task in task_group.tasks:
+                    t_obj, t_create = get_or_create(
+                        session=session, model=Tasks, name=task.entry)
+                    if task.meta_attr:
+                        t_obj.task_attr = json.dumps(task.meta_attr)
+                    if t_create:
+                        session.add(t_obj)
+
+                    sampel_obj = Samples(
+                        data_type=task.samples.data_type,
+                        sample_num=len(task.samples)
+                    )
+                    session.add(sampel_obj)
+
+                    session.flush()
+                    session.commit()
+                    sample_dir = FileOps.join_path(
+                        self.save_dir,
+                        f"{sampel_obj.data_type}_{sampel_obj.id}.pkl")
+                    task.samples.save(sample_dir)
+                    sampel_obj.data_url = sample_dir
+
+                    tsample = TaskSample(sample=sampel_obj, task=t_obj)
+                    session.add(tsample)
+                    session.flush()
+                    t_id.append(t_obj.id)
+
+                model_obj, m_create = get_or_create(
+                    session=session, model=TaskModel, task=grp)
+                model_obj.model_url = task_group.model.model
+                model_obj.is_current = False
+                if m_create:
+                    session.add(model_obj)
+                session.flush()
+                session.commit()
+                transfer_radio = 1 / grp.task_num
+                for t in t_id:
+                    t_obj, t_create = get_or_create(
+                        session=session, model=TaskRelation,
+                        task_id=t, grp=grp)
+                    t_obj.transfer_radio = transfer_radio
+                    if t_create:
+                        session.add(t_obj)
+                        session.flush()
+                    session.commit()
+                session.query(TaskRelation).filter(
+                    TaskRelation.grp == grp).update(
+                    {"transfer_radio": transfer_radio})
+
+            session.commit()
+
+        self.latest += 1
+        extractor_file = upload_info["extractor"]
+        extractor_path = FileOps.join_path(self.save_dir,
+                                           f"kb_extractor.pkl")
+        FileOps.upload(extractor_file, extractor_path)
+
+        # todo: get from kb
+        _index_path = FileOps.join_path(self.save_dir,
+                                        f"kb_index_{self.latest}.pkl")
+        FileOps.upload(name, _index_path)
+        res = f"/file/download?files=kb_index_{self.latest}.pkl&name=index.pkl"
+        return res
+
+
+if __name__ == '__main__':
+    KBServer(host="127.0.0.1").start()
```

Comment:
Is 127.0.0.1 right? listening to 127.0.0.1 will forbid train-worker access KB server

Reply:
fix